crates/aptos-logger/src/aptos_logger.rs

```diff
@@ -1024,4 +1055,135 @@ mod tests {
                 "source_path"
             )));
     }
+
+    // Test speculative logging via SpeculativeEvents, here because the testing infrastructure can be re-used.
+    struct SpeculativeLog {
+        level: Level,
+        message: String,
+    }
+
+    impl SpeculativeLog {
+        fn new(level: Level, message: String) -> Self {
+            Self { level, message }
+        }
+    }
+
+    impl SpeculativeEvent for SpeculativeLog {
+        fn dispatch(&self) {
+            match self.level {
+                Level::Debug => debug!("{}", self.message),
+                Level::Warn => warn!("{}", self.message),
+                Level::Error => error!("{}", self.message),
+                _ => {},
+            }
+        }
+    }
+
+    #[test]
+    fn test_speculative_logging() {
+        let receiver_mutex = set_test_logger(Level::Debug, true);
+        let receiver = receiver_mutex.lock();
+
+        while receiver.try_recv().is_ok() {}
+
+        let speculative_logs = SpeculativeEvents::<SpeculativeLog>::new(2);
+
+        // out of bounds
+        assert_err!(speculative_logs.record(
+            2,
+            SpeculativeLog::new(Level::Trace, "0/trace: A".to_string()),
+        ));
+
+        // level trace isn't enabled
+        assert_ok!(speculative_logs.record(
+            0,
+            SpeculativeLog::new(Level::Trace, "0/trace: A".to_string()),
+        ));
+
+        assert_ok!(speculative_logs.record(
+            0,
+            SpeculativeLog::new(Level::Debug, "0/debug: A".to_string()),
+        ));
+        assert_ok!(
+            speculative_logs.record(0, SpeculativeLog::new(Level::Info, "0/info: A".to_string()))
+        );
+        assert_ok!(
+            speculative_logs.record(0, SpeculativeLog::new(Level::Warn, "0/warn: A".to_string()))
+        );
+        assert_ok!(speculative_logs.record(
+            0,
+            SpeculativeLog::new(Level::Error, "/0error: A".to_string()),
+        ));
+        assert_ok!(speculative_logs.record(
+            1,
+            SpeculativeLog::new(Level::Error, "1/error: A".to_string()),
+        ));
+        assert_ok!(
+            speculative_logs.record(0, SpeculativeLog::new(Level::Warn, "0/warn: B".to_string()))
+        );
+        // Clear everything above.
```

Comment:
Part of me feels like this test can be split up into a couple different tests, each verifying different functionality.

Reply:
done :-) now the testing in the speculative crate is absolutely over-the-top, separate units, fake speculative logging tests and demo, and proptests LoL.
What would a man do to escape more important PRs.